### PR TITLE
Optimize Triton decoding kernel for long context

### PR DESCRIPTION
**Pull request by ispobock**
## Motivation

As mentioned in https://github.com/sgl-project/sglang/issues/2271, the original triton decoding kernel has significant performance degradation on long context. We refactored the kernel and adapted the flash decoding implementation from [lightllm](https://github.com/ModelTC/lightllm). Currently, the long context speed decay has been alleviated a lot.

## Benchmark
Tested for input 128, output 2048.

Triton (this PR) num_kv_splits=8: 150->138
```
$ python3 -m sglang.bench_offline_throughput --model meta-llama/Llama-3.1-8B-Instruct --dataset-name random --num-prompt 1 --random-input 128 --random-output 2048 --random-range 1 --attention-backend triton

[2024-12-08 08:37:24 TP0] Decode batch. #running-req: 1, #token: 194, token usage: 0.00, gen throughput (token/s): 150.66, #queue-req: 0

[2024-12-08 08:37:37 TP0] Decode batch. #running-req: 1, #token: 2154, token usage: 0.00, gen throughput (token/s): 138.86, #queue-req: 0
```

We can increase the `--triton-attention-num-kv-splits` to get better performance on long context.

Triton (this PR) num_kv_splits=16: 150->144
```
python3 -m sglang.bench_offline_throughput --model meta-llama/Llama-3.1-8B-Instruct --dataset-name random --num-prompt 1 --random-input 128 --random-output 2048 --random-range 1 --attention-backend triton --triton-attention-num-kv-splits 16

[2024-12-08 08:40:28 TP0] Decode batch. #running-req: 1, #token: 194, token usage: 0.00, gen throughput (token/s): 150.18, #queue-req: 0

[2024-12-08 08:40:42 TP0] Decode batch. #running-req: 1, #token: 2154, token usage: 0.00, gen throughput (token/s): 144.00, #queue-req: 0
```

Triton (main branch): 147->126
```
$ python3 -m sglang.bench_offline_throughput --model meta-llama/Llama-3.1-8B-Instruct --dataset-name random --num-prompt 1 --random-input 128 --random-output 2048 --random-range 1 --attention-backend triton

[2024-12-08 08:35:01 TP0] Decode batch. #running-req: 1, #token: 194, token usage: 0.00, gen throughput (token/s): 147.93, #queue-req: 0

[2024-12-08 08:35:15 TP0] Decode batch. #running-req: 1, #token: 2154, token usage: 0.00, gen throughput (token/s): 126.67, #queue-req: 0
```

Flashinfer: 143->143
```
$ python3 -m sglang.bench_offline_throughput --model meta-llama/Llama-3.1-8B-Instruct --dataset-name random --num-prompt 1 --random-input 128 --random-output 2048 --random-range 1

[2024-12-08 08:43:15 TP0] Decode batch. #running-req: 1, #token: 194, token usage: 0.00, gen throughput (token/s): 143.84, #queue-req: 0

[2024-12-08 08:43:29 TP0] Decode batch. #running-req: 1, #token: 2154, token usage: 0.00, gen throughput (token/s): 143.24, #queue-req: 0
```

### CODE REVIEW

File: python/sglang/srt/layers/attention/triton_ops/decode_attention.py [hunk 1]
```diff
@@ -705,10 +650,10 @@ def decode_attention_fwd(
             o,
             req_to_token,
             b_req_idx,
-            b_start_loc,
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/triton_ops/decode_attention.py:708]: remove this in the func signature of `decode_attention_fwd`?

File: python/sglang/srt/layers/attention/triton_backend.py [hunk 2]
```diff
@@ -53,10 +56,14 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
...
 
-            total_num_tokens = forward_batch.seq_lens_sum
             attn_logits = torch.empty(
-                (self.num_head, total_num_tokens),
-                dtype=self.reduce_dtype,
+                (
+                    forward_batch.batch_size,
+                    self.num_head,
+                    self.num_kv_splits,
+                    self.v_head_dim + 1,
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/triton_backend.py:64]: After this, we do not need to reduce the cuda graph max bs for deepseek models?

File: python/sglang/srt/layers/attention/triton_ops/decode_attention.py (hunk 1 above)
Comment (ispobock) [reply python/sglang/srt/layers/attention/triton_ops/decode_attention.py:708]: Sure. `max_len_in_batch` and `triton_attention_reduce_in_fp32` may also need to be removed.

File: python/sglang/srt/layers/attention/triton_backend.py (hunk 2 above)
Comment (ispobock) [reply python/sglang/srt/layers/attention/triton_backend.py:64]: Let me verify it.

**Review by merrymercy — APPROVED** [review]

**Comment by WANG-GH** [comment]
Hello, does issue #2394 still need further development? I saw this issue and would like to give it a try. However, I noticed that you are currently working on fixing it. Could you share the progress so far? Do you still need any help? @ispobock @merrymercy

**Comment by WANG-GH** [comment]
Hello, does issue #2271 still need further development? I saw this issue and would like to give it a try. However, I noticed that you are currently working on fixing it. Could you share the progress so far? Do you still need any help? @ispobock @merrymercy

**Pull request merged**